// azure-pipelines-tasks-packaging-common/universal/ArtifactToolRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runArtifactTool = exports.getOptions = void 0;
// TODO: Update functions in this file to call the implementations in ClientToolRunner.ts
const tl = require("azure-pipelines-task-lib");
let fs = require("fs");
let os = require("os");
const child = require("child_process");
function getOptions() {
    let result = {
        cwd: process.cwd(),
        env: Object.assign({}, process.env),
        silent: false,
        failOnStdErr: false,
        ignoreReturnCode: false,
        windowsVerbatimArguments: false
    };
    result.outStream = process.stdout;
    result.errStream = process.stderr;
    return result;
}
exports.getOptions = getOptions;
function getCommandString(toolPath, command) {
    let cmd = toolPath;
    command.forEach((a) => {
        cmd += ` ${a}`;
    });
    return cmd;
}
function runArtifactTool(artifactToolPath, command, execOptions) {
    if (tl.osType() === "Windows_NT" || artifactToolPath.trim().toLowerCase().endsWith(".exe")) {
        return tl.execSync(artifactToolPath, command, execOptions);
    }
    else {
        fs.chmodSync(artifactToolPath, "755");
        if (!execOptions.silent) {
            execOptions.outStream.write(getCommandString(artifactToolPath, command) + os.EOL);
        }
        let result = child.spawnSync(artifactToolPath, command, execOptions);
        if (!execOptions.silent && result.stdout && result.stdout.length > 0) {
            execOptions.outStream.write(result.stdout);
        }
        if (!execOptions.silent && result.stderr && result.stderr.length > 0) {
            execOptions.errStream.write(result.stderr);
        }
        let res = { code: result.status, error: result.error };
        res.stdout = (result.stdout) ? result.stdout.toString() : null;
        res.stderr = (result.stderr) ? result.stderr.toString() : null;
        return res;
    }
}
exports.runArtifactTool = runArtifactTool;
