// vso-node-api/interfaces/DistributedTaskCommonInterfaces.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
exports.TypeInfo = {
    DataSourceBindingBase: {
        fields: null
    },
    ProcessParameters: {
        fields: null
    },
    TaskInputDefinitionBase: {
        fields: null
    },
    TaskSourceDefinitionBase: {
        fields: null
    },
};
exports.TypeInfo.DataSourceBindingBase.fields = {};
exports.TypeInfo.ProcessParameters.fields = {
    dataSourceBindings: {
        isArray: true,
        typeInfo: exports.TypeInfo.DataSourceBindingBase
    },
    inputs: {
        isArray: true,
        typeInfo: exports.TypeInfo.TaskInputDefinitionBase
    },
    sourceDefinitions: {
        isArray: true,
        typeInfo: exports.TypeInfo.TaskSourceDefinitionBase
    },
};
exports.TypeInfo.TaskInputDefinitionBase.fields = {};
exports.TypeInfo.TaskSourceDefinitionBase.fields = {};
