// vso-node-api/interfaces/WorkItemTrackingInterfaces.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
const VSSInterfaces = require("../interfaces/common/VSSInterfaces");
var CommentSortOrder;
(function (CommentSortOrder) {
    CommentSortOrder[CommentSortOrder["Asc"] = 1] = "Asc";
    CommentSortOrder[CommentSortOrder["Desc"] = 2] = "Desc";
})(CommentSortOrder = exports.CommentSortOrder || (exports.CommentSortOrder = {}));
var FieldType;
(function (FieldType) {
    FieldType[FieldType["String"] = 0] = "String";
    FieldType[FieldType["Integer"] = 1] = "Integer";
    FieldType[FieldType["DateTime"] = 2] = "DateTime";
    FieldType[FieldType["PlainText"] = 3] = "PlainText";
    FieldType[FieldType["Html"] = 4] = "Html";
    FieldType[FieldType["TreePath"] = 5] = "TreePath";
    FieldType[FieldType["History"] = 6] = "History";
    FieldType[FieldType["Double"] = 7] = "Double";
    FieldType[FieldType["Guid"] = 8] = "Guid";
    FieldType[FieldType["Boolean"] = 9] = "Boolean";
})(FieldType = exports.FieldType || (exports.FieldType = {}));
var GetFieldsExpand;
(function (GetFieldsExpand) {
    GetFieldsExpand[GetFieldsExpand["None"] = 0] = "None";
    GetFieldsExpand[GetFieldsExpand["ExtensionFields"] = 1] = "ExtensionFields";
})(GetFieldsExpand = exports.GetFieldsExpand || (exports.GetFieldsExpand = {}));
var LinkChangeType;
(function (LinkChangeType) {
    LinkChangeType[LinkChangeType["Create"] = 0] = "Create";
    LinkChangeType[LinkChangeType["Remove"] = 1] = "Remove";
})(LinkChangeType = exports.LinkChangeType || (exports.LinkChangeType = {}));
var LinkQueryMode;
(function (LinkQueryMode) {
    LinkQueryMode[LinkQueryMode["WorkItems"] = 0] = "WorkItems";
    LinkQueryMode[LinkQueryMode["LinksOneHopMustContain"] = 1] = "LinksOneHopMustContain";
    LinkQueryMode[LinkQueryMode["LinksOneHopMayContain"] = 2] = "LinksOneHopMayContain";
    LinkQueryMode[LinkQueryMode["LinksOneHopDoesNotContain"] = 3] = "LinksOneHopDoesNotContain";
    LinkQueryMode[LinkQueryMode["LinksRecursiveMustContain"] = 4] = "LinksRecursiveMustContain";
    LinkQueryMode[LinkQueryMode["LinksRecursiveMayContain"] = 5] = "LinksRecursiveMayContain";
    LinkQueryMode[LinkQueryMode["LinksRecursiveDoesNotContain"] = 6] = "LinksRecursiveDoesNotContain";
})(LinkQueryMode = exports.LinkQueryMode || (exports.LinkQueryMode = {}));
var LogicalOperation;
(function (LogicalOperation) {
    LogicalOperation[LogicalOperation["NONE"] = 0] = "NONE";
    LogicalOperation[LogicalOperation["AND"] = 1] = "AND";
    LogicalOperation[LogicalOperation["OR"] = 2] = "OR";
})(LogicalOperation = exports.LogicalOperation || (exports.LogicalOperation = {}));
var ProvisioningActionType;
(function (ProvisioningActionType) {
    ProvisioningActionType[ProvisioningActionType["Import"] = 0] = "Import";
    ProvisioningActionType[ProvisioningActionType["Validate"] = 1] = "Validate";
})(ProvisioningActionType = exports.ProvisioningActionType || (exports.ProvisioningActionType = {}));
var QueryExpand;
(function (QueryExpand) {
    QueryExpand[QueryExpand["None"] = 0] = "None";
    QueryExpand[QueryExpand["Wiql"] = 1] = "Wiql";
    QueryExpand[QueryExpand["Clauses"] = 2] = "Clauses";
    QueryExpand[QueryExpand["All"] = 3] = "All";
})(QueryExpand = exports.QueryExpand || (exports.QueryExpand = {}));
var QueryOption;
(function (QueryOption) {
    QueryOption[QueryOption["Doing"] = 1] = "Doing";
    QueryOption[QueryOption["Done"] = 2] = "Done";
    QueryOption[QueryOption["Followed"] = 3] = "Followed";
})(QueryOption = exports.QueryOption || (exports.QueryOption = {}));
var QueryResultType;
(function (QueryResultType) {
    QueryResultType[QueryResultType["WorkItem"] = 1] = "WorkItem";
    QueryResultType[QueryResultType["WorkItemLink"] = 2] = "WorkItemLink";
})(QueryResultType = exports.QueryResultType || (exports.QueryResultType = {}));
var QueryType;
(function (QueryType) {
    QueryType[QueryType["Flat"] = 1] = "Flat";
    QueryType[QueryType["Tree"] = 2] = "Tree";
    QueryType[QueryType["OneHop"] = 3] = "OneHop";
})(QueryType = exports.QueryType || (exports.QueryType = {}));
var ReportingRevisionsExpand;
(function (ReportingRevisionsExpand) {
    ReportingRevisionsExpand[ReportingRevisionsExpand["None"] = 0] = "None";
    ReportingRevisionsExpand[ReportingRevisionsExpand["Fields"] = 1] = "Fields";
})(ReportingRevisionsExpand = exports.ReportingRevisionsExpand || (exports.ReportingRevisionsExpand = {}));
var TemplateType;
(function (TemplateType) {
    TemplateType[TemplateType["WorkItemType"] = 0] = "WorkItemType";
    TemplateType[TemplateType["GlobalWorkflow"] = 1] = "GlobalWorkflow";
})(TemplateType = exports.TemplateType || (exports.TemplateType = {}));
var TreeNodeStructureType;
(function (TreeNodeStructureType) {
    TreeNodeStructureType[TreeNodeStructureType["Area"] = 0] = "Area";
    TreeNodeStructureType[TreeNodeStructureType["Iteration"] = 1] = "Iteration";
})(TreeNodeStructureType = exports.TreeNodeStructureType || (exports.TreeNodeStructureType = {}));
var TreeStructureGroup;
(function (TreeStructureGroup) {
    TreeStructureGroup[TreeStructureGroup["Areas"] = 0] = "Areas";
    TreeStructureGroup[TreeStructureGroup["Iterations"] = 1] = "Iterations";
})(TreeStructureGroup = exports.TreeStructureGroup || (exports.TreeStructureGroup = {}));
var WorkItemErrorPolicy;
(function (WorkItemErrorPolicy) {
    WorkItemErrorPolicy[WorkItemErrorPolicy["Fail"] = 1] = "Fail";
    WorkItemErrorPolicy[WorkItemErrorPolicy["Omit"] = 2] = "Omit";
})(WorkItemErrorPolicy = exports.WorkItemErrorPolicy || (exports.WorkItemErrorPolicy = {}));
var WorkItemExpand;
(function (WorkItemExpand) {
    WorkItemExpand[WorkItemExpand["None"] = 0] = "None";
    WorkItemExpand[WorkItemExpand["Relations"] = 1] = "Relations";
    WorkItemExpand[WorkItemExpand["Fields"] = 2] = "Fields";
    WorkItemExpand[WorkItemExpand["Links"] = 3] = "Links";
    WorkItemExpand[WorkItemExpand["All"] = 4] = "All";
})(WorkItemExpand = exports.WorkItemExpand || (exports.WorkItemExpand = {}));
var WorkItemRecentActivityType;
(function (WorkItemRecentActivityType) {
    WorkItemRecentActivityType[WorkItemRecentActivityType["Visited"] = 0] = "Visited";
    WorkItemRecentActivityType[WorkItemRecentActivityType["Edited"] = 1] = "Edited";
    WorkItemRecentActivityType[WorkItemRecentActivityType["Deleted"] = 2] = "Deleted";
    WorkItemRecentActivityType[WorkItemRecentActivityType["Restored"] = 3] = "Restored";
})(WorkItemRecentActivityType = exports.WorkItemRecentActivityType || (exports.WorkItemRecentActivityType = {}));
exports.TypeInfo = {
    AccountMyWorkResult: {
        fields: null
    },
    AccountRecentActivityWorkItemModel: {
        fields: null
    },
    AccountWorkWorkItemModel: {
        fields: null
    },
    AttachmentReference: {
        fields: null
    },
    CommentSortOrder: {
        enumValues: {
            "asc": 1,
            "desc": 2,
        }
    },
    FieldDependentRule: {
        fields: null
    },
    FieldsToEvaluate: {
        fields: null
    },
    FieldType: {
        enumValues: {
            "string": 0,
            "integer": 1,
            "dateTime": 2,
            "plainText": 3,
            "html": 4,
            "treePath": 5,
            "history": 6,
            "double": 7,
            "guid": 8,
            "boolean": 9,
        }
    },
    GetFieldsExpand: {
        enumValues: {
            "none": 0,
            "extensionFields": 1,
        }
    },
    IdentityReference: {
        fields: null
    },
    Link: {
        fields: null
    },
    LinkChangeType: {
        enumValues: {
            "create": 0,
            "remove": 1,
        }
    },
    LinkQueryMode: {
        enumValues: {
            "workItems": 0,
            "linksOneHopMustContain": 1,
            "linksOneHopMayContain": 2,
            "linksOneHopDoesNotContain": 3,
            "linksRecursiveMustContain": 4,
            "linksRecursiveMayContain": 5,
            "linksRecursiveDoesNotContain": 6,
        }
    },
    LogicalOperation: {
        enumValues: {
            "nONE": 0,
            "aND": 1,
            "oR": 2,
        }
    },
    ProjectWorkItemStateColors: {
        fields: null
    },
    ProvisioningActionType: {
        enumValues: {
            "import": 0,
            "validate": 1,
        }
    },
    ProvisioningResult: {
        fields: null
    },
    QueryExpand: {
        enumValues: {
            "none": 0,
            "wiql": 1,
            "clauses": 2,
            "all": 3,
        }
    },
    QueryHierarchyItem: {
        fields: null
    },
    QueryOption: {
        enumValues: {
            "doing": 1,
            "done": 2,
            "followed": 3,
        }
    },
    QueryResultType: {
        enumValues: {
            "workItem": 1,
            "workItemLink": 2,
        }
    },
    QueryType: {
        enumValues: {
            "flat": 1,
            "tree": 2,
            "oneHop": 3,
        }
    },
    ReportingRevisionsExpand: {
        enumValues: {
            "none": 0,
            "fields": 1,
        }
    },
    ReportingWorkItemLink: {
        fields: null
    },
    ReportingWorkItemLinksBatch: {
        fields: null
    },
    ReportingWorkItemRevisionsBatch: {
        fields: null
    },
    ReportingWorkItemRevisionsFilter: {
        fields: null
    },
    StreamedBatch: {
        fields: null
    },
    TemplateType: {
        enumValues: {
            "workItemType": 0,
            "globalWorkflow": 1,
        }
    },
    TreeNodeStructureType: {
        enumValues: {
            "area": 0,
            "iteration": 1,
        }
    },
    TreeStructureGroup: {
        enumValues: {
            "areas": 0,
            "iterations": 1,
        }
    },
    Wiql: {
        fields: null
    },
    WorkItem: {
        fields: null
    },
    WorkItemClassificationNode: {
        fields: null
    },
    WorkItemComment: {
        fields: null
    },
    WorkItemComments: {
        fields: null
    },
    WorkItemDelete: {
        fields: null
    },
    WorkItemDeleteReference: {
        fields: null
    },
    WorkItemDeleteUpdate: {
        fields: null
    },
    WorkItemErrorPolicy: {
        enumValues: {
            "fail": 1,
            "omit": 2,
        }
    },
    WorkItemExpand: {
        enumValues: {
            "none": 0,
            "relations": 1,
            "fields": 2,
            "links": 3,
            "all": 4,
        }
    },
    WorkItemField: {
        fields: null
    },
    WorkItemFieldOperation: {
        fields: null
    },
    WorkItemFieldReference: {
        fields: null
    },
    WorkItemFieldUpdate: {
        fields: null
    },
    WorkItemHistory: {
        fields: null
    },
    WorkItemLink: {
        fields: null
    },
    WorkItemQueryClause: {
        fields: null
    },
    WorkItemQueryResult: {
        fields: null
    },
    WorkItemQuerySortColumn: {
        fields: null
    },
    WorkItemRecentActivityType: {
        enumValues: {
            "visited": 0,
            "edited": 1,
            "deleted": 2,
            "restored": 3,
        }
    },
    WorkItemReference: {
        fields: null
    },
    WorkItemRelation: {
        fields: null
    },
    WorkItemRelationType: {
        fields: null
    },
    WorkItemRelationUpdates: {
        fields: null
    },
    WorkItemStateColor: {
        fields: null
    },
    WorkItemStateTransition: {
        fields: null
    },
    WorkItemTemplate: {
        fields: null
    },
    WorkItemTemplateReference: {
        fields: null
    },
    WorkItemTrackingReference: {
        fields: null
    },
    WorkItemTrackingResource: {
        fields: null
    },
    WorkItemTrackingResourceReference: {
        fields: null
    },
    WorkItemType: {
        fields: null
    },
    WorkItemTypeCategory: {
        fields: null
    },
    WorkItemTypeColor: {
        fields: null
    },
    WorkItemTypeFieldInstance: {
        fields: null
    },
    WorkItemTypeReference: {
        fields: null
    },
    WorkItemTypeStateColors: {
        fields: null
    },
    WorkItemTypeTemplate: {
        fields: null
    },
    WorkItemTypeTemplateUpdateModel: {
        fields: null
    },
    WorkItemUpdate: {
        fields: null
    },
};
exports.TypeInfo.AccountMyWorkResult.fields = {
    workItemDetails: {
        isArray: true,
        typeInfo: exports.TypeInfo.AccountWorkWorkItemModel
    },
};
exports.TypeInfo.AccountRecentActivityWorkItemModel.fields = {
    activityDate: {
        isDate: true,
    },
    activityType: {
        enumType: exports.TypeInfo.WorkItemRecentActivityType
    },
    changedDate: {
        isDate: true,
    },
};
exports.TypeInfo.AccountWorkWorkItemModel.fields = {
    changedDate: {
        isDate: true,
    },
};
exports.TypeInfo.AttachmentReference.fields = {};
exports.TypeInfo.FieldDependentRule.fields = {
    dependentFields: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
};
exports.TypeInfo.FieldsToEvaluate.fields = {};
exports.TypeInfo.IdentityReference.fields = {};
exports.TypeInfo.Link.fields = {};
exports.TypeInfo.ProjectWorkItemStateColors.fields = {
    workItemTypeStateColors: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemTypeStateColors
    },
};
exports.TypeInfo.ProvisioningResult.fields = {};
exports.TypeInfo.QueryHierarchyItem.fields = {
    children: {
        isArray: true,
        typeInfo: exports.TypeInfo.QueryHierarchyItem
    },
    clauses: {
        typeInfo: exports.TypeInfo.WorkItemQueryClause
    },
    columns: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
    createdBy: {
        typeInfo: exports.TypeInfo.IdentityReference
    },
    createdDate: {
        isDate: true,
    },
    filterOptions: {
        enumType: exports.TypeInfo.LinkQueryMode
    },
    lastModifiedBy: {
        typeInfo: exports.TypeInfo.IdentityReference
    },
    lastModifiedDate: {
        isDate: true,
    },
    linkClauses: {
        typeInfo: exports.TypeInfo.WorkItemQueryClause
    },
    queryType: {
        enumType: exports.TypeInfo.QueryType
    },
    sortColumns: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemQuerySortColumn
    },
    sourceClauses: {
        typeInfo: exports.TypeInfo.WorkItemQueryClause
    },
    targetClauses: {
        typeInfo: exports.TypeInfo.WorkItemQueryClause
    },
};
exports.TypeInfo.ReportingWorkItemLink.fields = {
    changedBy: {
        typeInfo: VSSInterfaces.TypeInfo.IdentityRef
    },
    changedDate: {
        isDate: true,
    },
    changedOperation: {
        enumType: exports.TypeInfo.LinkChangeType
    },
};
exports.TypeInfo.ReportingWorkItemLinksBatch.fields = {};
exports.TypeInfo.ReportingWorkItemRevisionsBatch.fields = {};
exports.TypeInfo.ReportingWorkItemRevisionsFilter.fields = {};
exports.TypeInfo.StreamedBatch.fields = {};
exports.TypeInfo.Wiql.fields = {};
exports.TypeInfo.WorkItem.fields = {
    relations: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemRelation
    },
};
exports.TypeInfo.WorkItemClassificationNode.fields = {
    children: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemClassificationNode
    },
    structureType: {
        enumType: exports.TypeInfo.TreeNodeStructureType
    },
};
exports.TypeInfo.WorkItemComment.fields = {
    revisedBy: {
        typeInfo: exports.TypeInfo.IdentityReference
    },
    revisedDate: {
        isDate: true,
    },
};
exports.TypeInfo.WorkItemComments.fields = {
    comments: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemComment
    },
};
exports.TypeInfo.WorkItemDelete.fields = {
    resource: {
        typeInfo: exports.TypeInfo.WorkItem
    },
};
exports.TypeInfo.WorkItemDeleteReference.fields = {};
exports.TypeInfo.WorkItemDeleteUpdate.fields = {};
exports.TypeInfo.WorkItemField.fields = {
    supportedOperations: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemFieldOperation
    },
    type: {
        enumType: exports.TypeInfo.FieldType
    },
};
exports.TypeInfo.WorkItemFieldOperation.fields = {};
exports.TypeInfo.WorkItemFieldReference.fields = {};
exports.TypeInfo.WorkItemFieldUpdate.fields = {};
exports.TypeInfo.WorkItemHistory.fields = {
    revisedBy: {
        typeInfo: exports.TypeInfo.IdentityReference
    },
    revisedDate: {
        isDate: true,
    },
};
exports.TypeInfo.WorkItemLink.fields = {
    source: {
        typeInfo: exports.TypeInfo.WorkItemReference
    },
    target: {
        typeInfo: exports.TypeInfo.WorkItemReference
    },
};
exports.TypeInfo.WorkItemQueryClause.fields = {
    clauses: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemQueryClause
    },
    field: {
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
    fieldValue: {
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
    logicalOperator: {
        enumType: exports.TypeInfo.LogicalOperation
    },
    operator: {
        typeInfo: exports.TypeInfo.WorkItemFieldOperation
    },
};
exports.TypeInfo.WorkItemQueryResult.fields = {
    asOf: {
        isDate: true,
    },
    columns: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
    queryResultType: {
        enumType: exports.TypeInfo.QueryResultType
    },
    queryType: {
        enumType: exports.TypeInfo.QueryType
    },
    sortColumns: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemQuerySortColumn
    },
    workItemRelations: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemLink
    },
    workItems: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemReference
    },
};
exports.TypeInfo.WorkItemQuerySortColumn.fields = {
    field: {
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
};
exports.TypeInfo.WorkItemReference.fields = {};
exports.TypeInfo.WorkItemRelation.fields = {};
exports.TypeInfo.WorkItemRelationType.fields = {};
exports.TypeInfo.WorkItemRelationUpdates.fields = {
    added: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemRelation
    },
    removed: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemRelation
    },
    updated: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemRelation
    },
};
exports.TypeInfo.WorkItemStateColor.fields = {};
exports.TypeInfo.WorkItemStateTransition.fields = {};
exports.TypeInfo.WorkItemTemplate.fields = {};
exports.TypeInfo.WorkItemTemplateReference.fields = {};
exports.TypeInfo.WorkItemTrackingReference.fields = {};
exports.TypeInfo.WorkItemTrackingResource.fields = {};
exports.TypeInfo.WorkItemTrackingResourceReference.fields = {};
exports.TypeInfo.WorkItemType.fields = {
    fieldInstances: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemTypeFieldInstance
    },
    fields: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemTypeFieldInstance
    },
    transitions: {},
};
exports.TypeInfo.WorkItemTypeCategory.fields = {
    defaultWorkItemType: {
        typeInfo: exports.TypeInfo.WorkItemTypeReference
    },
    workItemTypes: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemTypeReference
    },
};
exports.TypeInfo.WorkItemTypeColor.fields = {};
exports.TypeInfo.WorkItemTypeFieldInstance.fields = {
    field: {
        typeInfo: exports.TypeInfo.WorkItemFieldReference
    },
};
exports.TypeInfo.WorkItemTypeReference.fields = {};
exports.TypeInfo.WorkItemTypeStateColors.fields = {
    stateColors: {
        isArray: true,
        typeInfo: exports.TypeInfo.WorkItemStateColor
    },
};
exports.TypeInfo.WorkItemTypeTemplate.fields = {};
exports.TypeInfo.WorkItemTypeTemplateUpdateModel.fields = {
    actionType: {
        enumType: exports.TypeInfo.ProvisioningActionType
    },
    templateType: {
        enumType: exports.TypeInfo.TemplateType
    },
};
exports.TypeInfo.WorkItemUpdate.fields = {
    fields: {},
    relations: {
        typeInfo: exports.TypeInfo.WorkItemRelationUpdates
    },
    revisedBy: {
        typeInfo: exports.TypeInfo.IdentityReference
    },
    revisedDate: {
        isDate: true,
    },
};
