// azure-pipelines-tasks-packaging-common/pe-parser/VersionInfoVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionInfoVersion = void 0;
class VersionInfoVersion {
    constructor(a, b, c, d) {
        this.a = a;
        this.b = b;
        this.c = c;
        this.d = d;
    }
    static fromDWords(highDWord, lowDWord) {
        // tslint:disable-next-line:no-bitwise
        return new VersionInfoVersion(highDWord >>> 16, highDWord & 0xFFFF, lowDWord >>> 16, lowDWord & 0xFFFF);
    }
    static compare(left, right) {
        return left.a - right.a || left.b - right.b || left.c - right.c || left.d - right.d;
    }
    toString() {
        return `${this.a}.${this.b}.${this.c}.${this.d}`;
    }
    equals(other) {
        return VersionInfoVersion.compare(this, other) === 0;
    }
}
exports.VersionInfoVersion = VersionInfoVersion;
VersionInfoVersion.MAX_VERSION = new VersionInfoVersion(0xFFFF, 0xFFFF, 0xFFFF, 0xFFFF);
VersionInfoVersion.MIN_VERSION = new VersionInfoVersion(0x0, 0x0, 0x0, 0x0);
exports.default = VersionInfoVersion;
