// azure-pipelines-tasks-packaging-common/pe-parser/ReadableFile.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadableFile = void 0;
const fs = require("fs");
const Q = require("q");
class ReadableFile {
    constructor(path, fd) {
        this.path = path;
        this.fd = fd;
    }
    static openAsync(path) {
        return __awaiter(this, void 0, void 0, function* () {
            let fd = yield Q.nfcall(fs.open, path, "r");
            return new ReadableFile(path, fd);
        });
    }
    readAsync(buffer, offset, length, position) {
        return __awaiter(this, void 0, void 0, function* () {
            return Q.nfcall(fs.read, this.fd, buffer, offset, length, position);
        });
    }
    closeAsync() {
        return __awaiter(this, void 0, void 0, function* () {
            return Q.nfcall(fs.close, this.fd);
        });
    }
}
exports.ReadableFile = ReadableFile;
exports.default = ReadableFile;
